satellite/metainfo/metainfo.go

```diff
@@ -1247,9 +1223,55 @@ func (endpoint *Endpoint) BeginSegment(ctx context.Context, req *pb.SegmentBegin
 		return nil, status.Errorf(codes.Unauthenticated, err.Error())
 	}
 
-	// TODO implement logic
+	// no need to validate streamID fields because it was validated during BeginObject
+
+	exceeded, limit, err := endpoint.projectUsage.ExceedsStorageUsage(ctx, keyInfo.ProjectID)
+	if err != nil {
+		endpoint.log.Error("retrieving project storage totals", zap.Error(err))
+	}
+	if exceeded {
+		endpoint.log.Sugar().Errorf("monthly project limits are %s of storage and bandwidth usage. This limit has been exceeded for storage for projectID %s",
+			limit, keyInfo.ProjectID,
+		)
+		return nil, status.Errorf(codes.ResourceExhausted, "Exceeded Usage Limit")
+	}
+
+	redundancy, err := eestream.NewRedundancyStrategyFromProto(streamID.Redundancy)
+	if err != nil {
+		return nil, err
+	}
+
+	maxPieceSize := eestream.CalcPieceSize(req.MaxOrderLimit, redundancy)
 
-	return &pb.SegmentBeginResponse{}, status.Error(codes.Unimplemented, "not implemented")
+	request := overlay.FindStorageNodesRequest{
+		RequestedCount: redundancy.TotalCount(),
+		FreeBandwidth:  maxPieceSize,
+		FreeDisk:       maxPieceSize,
+	}
+	nodes, err := endpoint.cache.FindStorageNodes(ctx, request)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	bucketID := createBucketID(keyInfo.ProjectID, streamID.Bucket)
+	rootPieceID, addressedLimits, piecePrivateKey, err := endpoint.orders.CreatePutOrderLimits(ctx, bucketID, nodes, streamID.ExpirationDate, maxPieceSize)
+	if err != nil {
+		return nil, Error.Wrap(err)
```

Comment:
should this be another status `internal` error?

Reply:
fixed